torch/utils/mkldnn.py

```diff
@@ -27,7 +27,10 @@ def __setstate__(self, state):
 
     @torch.jit.script_method
     def forward(self, x):
-        return torch._C._nn.mkldnn_linear(x, self.weight, self.bias)
+        x_mkldnn = x if x.is_mkldnn else x.to_mkldnn()
+        y_mkldnn = torch._C._nn.mkldnn_linear(x_mkldnn, self.weight, self.bias)
+        y = y_mkldnn if x.is_mkldnn else y_mkldnn.to_dense()
+        return y
```

Comment:
@mingfeima We also need to put the similar logic to MkldnnConv2d and MkldnnBatchNorm2d, right?

Reply:
presumably... by hooking conversion logic into utils gave us more freedom... maybe we can try some fancy ideas that is impossible before... e.g. leading dimension padding perhaps... aka always use the best config for each input config.
i will do some experiments on this, might give some solid advantage over other FWs